atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -162,27 +168,40 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
     }
 
     @Override
-    public TransactionAndImmutableTsLock setupRunTaskWithConditionThrowOnConflict(PreCommitCondition condition) {
-        StartIdentifiedAtlasDbTransactionResponse transactionResponse
-                = timelockService.startIdentifiedAtlasDbTransaction();
+    public List<TransactionAndImmutableTsLock> setupRunTaskBatchWithConditionThrowOnConflict(
+            List<PreCommitCondition> conditions) {
+        List<StartIdentifiedAtlasDbTransactionResponse> responses =
+                timelockService.startIdentifiedAtlasDbTransactionBatch(conditions.size());
         try {
-            LockToken immutableTsLock = transactionResponse.immutableTimestamp().getLock();
-            long immutableTs = transactionResponse.immutableTimestamp().getImmutableTimestamp();
+            long immutableTs = Collections.max(responses.stream()
+                    .map(response -> response.immutableTimestamp().getImmutableTimestamp())
+                    .collect(Collectors.toList()));
             recordImmutableTimestamp(immutableTs);
 
-            cleaner.punch(transactionResponse.startTimestampAndPartition().timestamp());
-            Supplier<Long> startTimestampSupplier = Suppliers.ofInstance(
-                    transactionResponse.startTimestampAndPartition().timestamp());
-
-            Transaction transaction = createTransaction(
-                    immutableTs,
-                    startTimestampSupplier,
-                    immutableTsLock,
-                    condition);
-            return TransactionAndImmutableTsLock.of(transaction, immutableTsLock);
-        } catch (Throwable e) {
-            timelockService.tryUnlock(ImmutableSet.of(transactionResponse.immutableTimestamp().getLock()));
-            throw Throwables.rewrapAndThrowUncheckedException(e);
+            return Streams.zip(
+                    responses.stream(),
+                    conditions.stream(),
+                    (response, condition) -> {
+                        LockToken immutableTsLock = response.immutableTimestamp().getLock();
+                        cleaner.punch(response.startTimestampAndPartition().timestamp());
```

Comment:
Haven't looked at puncher in a while, so unclear if you can do this once for the whole batch same as immutable timestamp.

Reply:
From the javadoc:
```
Indicate that the given timestamp has just been created. This must be called frequently (preferably on each transaction commit) so that the Cleaner can keep track of the wall-clock/timestamp mapping. If it is never called, semantically nothing goes wrong, but the sweeper won't sweep, since it cannot know what things are old enough to be swept.
```
I interpret this as something we could conceivably do once per batch. @jeremyk-91 to sanity check that.